Optimize InstanceNormGradientOp

xiaomengy opened the pull request:
Summary: Optimize InstanceNormGradientOp

Differential Revision: D16023630

BIT-silence commented:
Benchmarks:

CPU with [N, C, H, W] = [128, 256, 56, 56],
NCHW order: 616ms -> 128ms
NHWC order: 1612ms -> 174ms

GPU with [N, C, H, W] = [128, 256, 112, 112],
NCHW order: 6450ms -> 37ms
NHWC order: 1419ms -> 82ms

houseroad commented on caffe2/python/operator_test/instance_norm_test.py:
```diff
@@ -76,7 +76,16 @@ def test_instance_norm_gradients(
             order=order)
-        input_blobs = self._get_inputs(N, C, H, W, order)
+
+        input_data = np.arange(N * C * H * W).astype(np.float32)
```
why moved away from `_get_inputs`?

BIT-silence replied:
_get_inputs used rand to generate input, whose range is pretty small and may lead to numerical failures for grad check. Here we use arrange and shuffle to generate a tensor those elements are far away from each other.

houseroad replied:
I see, shall we just extend _get_inputs a bit? Since most of the code looks very similar

BIT-silence replied:
The correct thing is just remove _get_inputs at all. There is no need for 3 lines in such a function.

houseroad commented on caffe2/operators/instance_norm_op.h:
```diff
@@ -127,44 +127,148 @@ class InstanceNormOp final : public Operator<Context> {
+    const T* X_data = X.template data<T>();
+    const T* gamma_data = gamma.template data<T>();
+    const T* mean_data = nullptr;
+    const T* rstd_data = nullptr;
```
Also check the input number?

BIT-silence replied:
Done.

houseroad reviewed (approved):
Looks good.

xiaomengy committed:
Optimize InstanceNormGradientOp (#22288)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/22288

Optimize InstanceNormGradientOp

Benchmarks:

CPU with [N, C, H, W] = [128, 256, 56, 56],
NCHW order: 616ms -> 128ms
NHWC order: 1612ms -> 174ms

GPU with [N, C, H, W] = [128, 256, 112, 112],
NCHW order: 6450ms -> 37ms
NHWC order: 1419ms -> 82ms

Reviewed By: houseroad

Differential Revision: D16023630

fbshipit-source-id: f66100dbe5d86a8804843ade1915edba3a24a9eb
